### ISSUE TITLE
Completions for string literal don't work for type parameter that includes empty string in it's constraint 

### ISSUE THREAD

**Issue by devanshj** [microsoft/TypeScript#47227]
# Bug Report

### 🔎 Search Terms

String literal competitions, empty string constraint, intellisense

### 🕗 Version & Regression Information

Tested with v4.5.4

### ⏯ Playground Link

[Playground](https://tsplay.dev/wQevZw)

### 💻 Code

```ts
declare const foo1: (p: "" | "bar" | "baz") => void

foo1("")
//    ^|
// expected completions: "", "bar", "baz"
// actual completions: "", "bar", "baz"

declare const foo2: <P extends "" | "bar" | "baz">(p: P) => void

foo2("")
//    ^|
// expected completions: "", "bar", "baz"
// actual completions: none
```

### 🙁 Actual behavior

`foo2` has no completions

### 🙂 Expected behavior

`foo2` should have same completions like `foo1` which are `""`, `"bar"`, `"baz"`

**Comment by david-shortman** [comment]
When typing a `"` in the `foo1`, I don't receive suggestions until I type a `b` character.

<img width="521" alt="Screen Shot 2021-12-22 at 20 29 15" src="https://user-images.githubusercontent.com/25187726/147178723-6ce9a88d-ffaf-4ff6-8f9c-961eb5222c73.png">

When I type a `b` character in a string given to `foo2`, I get two suggestions.

<img width="572" alt="Screen Shot 2021-12-22 at 20 28 57" src="https://user-images.githubusercontent.com/25187726/147178722-478f3025-5174-4e55-82dc-907c13c43b36.png">

**Comment by devanshj** [comment]
Yes that's because `""` is a valid value so you don't get suggested anything once you're done writing it. But you can trigger intellisense via <kbd>CTRL</kbd> + <kbd>SPACE</kbd> while caret is between the quotes (like `"|"` then hit <kbd>CTRL</kbd> + <kbd>SPACE</kbd>). And then `foo1` will have suggestions and `foo2` won't:

![](https://user-images.githubusercontent.com/30295578/147180343-c9a5c297-7bf1-42f5-abca-9915272917fb.png)

![](https://user-images.githubusercontent.com/30295578/147180394-f050726f-d8a7-45bd-be64-efe6057f83b3.png)

Sometimes you'd want to look up all possible values before you even start typing something, which you can't do in case of `foo2`

**Comment by david-shortman** [comment]
The issue is specifically with empty strings, it seems. I tried using the suggestion shortcut in an empty argument list, and was presented with the three possible string options for both `foo1` and `foo2`.

<img width="505" alt="Screen Shot 2021-12-22 at 20 51 43" src="https://user-images.githubusercontent.com/25187726/147180856-0ca55266-0675-4ea3-941f-7de14e2dbe4f.png">

<img width="542" alt="Screen Shot 2021-12-22 at 20 51 53" src="https://user-images.githubusercontent.com/25187726/147180854-c39dbecd-467f-4945-9d59-9b4340cc6f84.png">

**Comment by devanshj** [comment]
Ah, yeah then it seems the issue occurres only when you have already written `""`

**Comment by Andarist** [comment]
This isn't really about an empty string but about the string literal at the argument position already matching one of the union members. The same happens here:
```ts
declare const foo2: <P extends "bar" | "barbaz">(p: P) => void

foo2('bar') // we won't get autocomplete for 'barbaz'
```

**Comment by Andarist** [comment]
Ok, so the "problem" is that autocomplete here works on the "resolved signature" and since the function is generic this already includes an inferred generic:
```ts
declare const foo2: <P extends "bar" | "barbaz">(p: P) => void

// signature: function foo2<"bar">(p: "bar"): void
foo2('bar')
```

This can be verified~ when hovering over the function symbol at the call site.

When the generic is not properly inferred because we have a signature applicability error then the signature is displayed like this:
```ts
declare const foo2: <P extends "bar" | "barbaz">(p: P) => void

// signature: const foo2: <"bar" | "barbaz">(p: "bar" | "barbaz") => void
foo2('unknown')
```

and that is what makes it work~.

I've also looked into how this behaves with overloads and we get a full list of completions until we match one of the signatures (so basically it's very similar to single-signature functions discussed above):
```ts
declare function bar1<P extends "bar" | "barbaz">(p: P): number;
declare function bar1<P extends "qwe" | "qwert">(p: P): string;

// completions: "bar" | "barbaz" | "qwe" | "qwert"
bar1('') 
```

This is thanks to the fact that `getResolvedSignatureForStringLiteralCompletions` aggregates all candidate signatures in the `candidates` array (after all nothing has been matched, so all signatures are still candidates).

Some other interesting bits:
- the `resolveCall`/`chooseOverload` functions are called within `runWithInferenceBlockedFromSourceNode` but this doesn't block signature selection
- this happens with `CheckMode.IsForStringLiteralArgumentCompletions` and that potentially could be used to `resolveCall` differently here

@andrewbranch do you see any specific implementation difficulties for this? Or a rather simple heuristic could be used here? I would have to dive deeper into how **different** positions within the arguments list are resolved, potentially they might need the `chooseOverload` calls to narrow down candidates and this might be quite problematic. Ideally, we would skip inference for the particular position that is being the subject of the autocomplete while narrowing down based on other arguments etc. Since the same generic can appear at different positions this probably gets complicated quite quickly 😬

**Comment by andrewbranch** [comment]
Isn’t this exactly what I fixed at https://github.com/microsoft/TypeScript/pull/48410? I’m guessing not, since you’re referring to methods and CheckModes I added there. But what you’re describing is what I remember doing 😅

**Comment by Andarist** [comment]
Ye, this is very related - the difference is that in [the test case that you have added](https://github.com/microsoft/TypeScript/blob/71f94c5bdb806b1dd835cf5a14bc56413bd62cbf/tests/cases/fourslash/completionsLiteralOverload.ts#L15-L16) there is no overload selected because there is a mismatch between the arity of the function declaration and the supplied arguments count. So the added logic from your PR (https://github.com/microsoft/TypeScript/pull/48410) handles that situation within `inferTypeArguments` that is called (transitively) from `getCandidateForOverloadFailure`. In this situation here - there is no overload failure though.
